libdnf/system/state.hpp

```diff
@@ -223,6 +228,9 @@ class State {
     // map caches groups for packages. Package name is the key, set of group ids is the value
     const std::map<std::string, std::set<std::string>> & get_package_groups_cache();
     std::optional<std::map<std::string, std::set<std::string>>> package_groups_cache;
+    // map caches packages by the reason. The key is reason (as string), value set of NA
+    const std::map<std::string, std::set<std::string>> & get_packages_by_reason_cache();
+    std::optional<std::map<std::string, std::set<std::string>>> packages_by_reason_cache;
```

Comment:
Is the cache really needed? Maybe it would be enough to just calculate it every time, assuming this will just be used for a potential command and not frequently in code in general. Though the cache probably doesn't hurt anything.

Reply:
For `group remove` usecase I need to ask for every package installed with the group, so I consider the `package_groups_cache` useful. I'm not that convinced about the `package_by_reason_cache` though.

Reply:
Do you prefer to remove the cache? Actually the `get_packages_by_reason()` function is called twice, but once with `GROUP` argument where this cache does not apply... Hm, maybe it should be called third time with `EXTERNAL_USER` reason. But the API can still be changed to accept a list of reasons...

Reply:
I don't see the calls, are they in the other PR?

I guess you can keep it, apart from the small memory consumption it doesn't hurt anything, just adding some complexity which I'm not sure is warrantied.

Reply:
I've added a commit with cache removal. I can squash them if you like it.
Yes, the calls are in other PR. Here is the draft - https://github.com/rpm-software-management/dnf5/pull/35. It needs to be adjusted to the latest changes.

Reply:
Yeah, if you remove it, please squash, thanks!

Reply:
Removed and squashed.